### DISCUSSION TITLE
Notebook Documentation Known Issues and Expectations

### DISCUSSION THREAD

**Discussion by kartikmodi** [radical-cybertools/radical.pilot#2537] [Q&A] [answered]
Hello, starting off this discussion to list down some minor known issues in notebook documentation (observed on my setup) and expectations from it. 

Status: I am able to run a basic RP session of /bin/date. My knowledge is extremely limited at this point. 
I am able to generate RST documentation as well as ipynb documentation from sphinx.

1. **MEDIUM** - The notebook doesn't run automatically when we built the documentation. Currently, I run the entire notebook separately and generate the documentation separately. Ofcourse we can run both as commands, one after the other, but is the right expectations or I am missing something. 

I imagined as when I build the documentation with sphinx it should run the notebooks then and there itself, before publishing documentation in HTML. 
> Need team input on this. 

2. **LOW** - Output problems
From notebook in the browser (jupyter UI) - 
```wait for 1 pilot(s)
        O39m39m39m39m39m39m39m39m39m39m39m39m39m39m39m                                                               ok
```
From HTML published by sphinx - 
```wait for 1 pilot(s)
        O/-\|/-\|      0                                                               ok
                                                                              ok
+ rp.session.debian-x.workstation.019053.0012 (json)
```
From embedded notebook inside a popular IDE (my preference)- 
```
Really BAD - ALL # are 039
```
> No inputs necessary as of now. 

3. **LOW** - `__file__` is not available in python (interactive shell) - 
We read configuration with taking reference of current file (current directory) to read resource. The `__file__` is not available so need some solution/workaround/better approach. 
I'll work on this soon.  
> No inputs necessary as of now. 
4. The documentation generated by sphinx is not interactive. That means the notebook contents and its output from cells are just directly embedded in the published HTML, that's it. It is not live/interactive or any sort of `Try it yourself thing. I know it's weird to expect that, but I am writing here just to clarify.  
> No inputs necessary as of now. 

5. **MEDIUM** - Loading env variables from a known file (.env default filename) - 
Since we need to run NBs I need to pass the env variables to notebooks. 
I usually use a .env file (pretty common practice) consisting of key-value variables. I use this  file because it's common to find developer tools like IDE plugins,  vim plugins which read this file, before launching the program. 
For notebooks however, I found one pypy package to use which works with like - 
```
%load_ext dotenv 
%dotenv
```
This loads the extension (py package) and then variables. I need ideas on how to handle env variables in NBs. This needs to be standardized as it will be in the build process and will be run every release/PR merge. 

6. **LOW** - List of dependencies installed until now to make this work - 
pip install nbsphinx
pip install sphinx_copybutton
pip install sphinx_rtd_theme
Need to add these in setup.py properly, don't know how. These cannot be installed in default installation. 
Maybe tests require could be a option. Discuss. 
> Need inputs here.

7. There are more issues/highlights, will add them soon.

I'll open a discussion separately if any of the above topics gets big enough, for better management.

**Comment by mturilli** [comment, marked as answer]
Hi @kartikmodi, here some initial notes, happy to discuss them more in detail on a dedicated meeting:
1. Yes, the expectation would be to execute the notebook via readthedocs. I tried that myself and, ultimately, I was not successful. I can share the way in which I addressed the issue of using an external MDB with authentication but more work needs to be done to understand how to debug RP when running it into a container without ssh access. I would be happy for the two of us to work together on that.
2. I noticed the same. Those look like control sequences that are not interpreted within a notebook cell. If so, I would avoid to assume that RP runs only on a POSIX sh environment.
3. I have the same problem. I asked the devel team to fix that few months back but still no joy.
4. Yes, we do not assume interactivity as readthedocs is a static documentation env.
5. That might be a way. Another one is to set those variables in readthedocs, either as secret of public values. I did  that for MDB. There may be other approaches, happy to go with the one you find more commonly used and useful.
6. If it is only in the context of the documentation, we can use the dependency file of Sphinx/readthedocs.

**Comment by kartikmodi** [comment]
Rethinking notebook documentation problem - 
Objectives - 

1. Is the objective to automate the building of the documentation as tests, during the build process, so that documentation stays valid, up-to-date. Failing to build the documentation would fail the release. 

2. Re-writing / Fixing parts of documentation because some parts are out-of-date/invalid, some miss key details, improvements for first-time readers, etc. 

3. Migrating to notebook-style writing i.e. code and markdown for its ease, modern adoption. 

The current Approach we are following solves all 3 stated objectives. 

---------------------------------------------------------------------------------------------------------------------------------------------------------------------
However, a relatively tweaked approach will solve Objective 1 and Objective 2, with a fraction of the time and cost. See here - 
https://github.com/radical-cybertools/radical.pilot/blob/docs/nb_section3/docs/source/3/3.3-handle-failing-tasks.ipynb
Check the Raw as well. 
https://raw.githubusercontent.com/radical-cybertools/radical.pilot/docs/nb_section3/docs/source/3/3.3-handle-failing-tasks.ipynb

Essentially I am making a wrapper notebook with 3 cells only - 
1. .rst doc inclusion 
2. env variable loading
3. .py example code inclusion

One consideration of Objective 3 is, we will move away from .rst in favour of .md. In my limited exposure to rst and md, this seems to be re-considered and discussed. 
@mturilli @andre-merzky

**Comment by eirrgang** [comment]
It seems like the intention is to build this documentation on readthedocs, but I'm not seeing a `.readthedocs.yml` file (which would be appropriate, if not necessary, to support more advanced configuration, such as non-default config file names/locations).

I don't think that custom environment variables are part of the rtd configuration schema, but they can be [set through the web interface](https://docs.readthedocs.io/en/stable/environment-variables.html#environment-variables). Alternatively, you could check for the READTHEDOCS environment variable to be set, and react within the module to choose appropriate default values.

I've had trouble with virtual environments on readthedocs in the past. It could be tricky to automate rtd to run notebooks that include creating sessions.

In other words, it might be easier to migrate to a basic scripted workflow on GitHub Actions.
